src/transformers/utils/generic.py

```diff
@@ -263,7 +263,28 @@ def __init_subclass__(cls) -> None:
                 lambda values, context: cls(**torch.utils._pytree._dict_unflatten(values, context)),
             )
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Subclasses of ModelOutput must use the @dataclass decorator
+        # This check is done in __init__ because the @dataclass decorator operates after __init_subclass__
+        # issubclass() would return True for issubclass(ModelOutput, ModelOutput) when False is needed
+        # Just need to check that the current class is not ModelOutput
+        is_modeloutput_subclass = self.__class__ != ModelOutput
+
+        if is_modeloutput_subclass and not is_dataclass(self):
+            raise TypeError(
+                (
+                    f"{self.__module__}.{self.__class__.__name__} is not a dataclasss."
+                    " This is a subclass of ModelOutput and so must use the @dataclass decorator."
+                )
+            )
```

Comment:
nit: extra brackets aren't needed here 

```suggestion
            raise TypeError(
                f"{self.__module__}.{self.__class__.__name__} is not a dataclasss."
                " This is a subclass of ModelOutput and so must use the @dataclass decorator."
            )
```

Reply:
Thanks!